### PR TITLE
Add `action_dann_lightn` to Pykale

### PR DESCRIPTION
**Pull request by XianyuanLiu**
### 1. Overview
I have followed the `digit_dann_lightn` to create a new example `action_dann_lightn` for action recognition domain adaptation.

### 2. Description
Main work is on the feature extractors and data loaders. I have tried the best to imitate the original `digit_dann_lightn` structures.

Video feature extractor networks (Res3D, I3D) are in `kale.embed.video_cnn`, while Data loaders for video datasets are in `kale.loaddata.video_access`. Meanwhile, in `kale.loaddata.video_data`, `basic_video_dataset` is for loading data from GTEA, KITCHEN and ADL Datasets and `epickitchen` is inherited from `basic_video_dataset` for EPIC-Kitchen dataset. `kale.prepdata.video_transform` contains transforms for video data. 

I have also added the docstring for all new codes and update the README.md and requirements.txt in `action_dann_lightn`.

### 3. TODO
* Modify the connection between feature extractor and classifier. (Too many linear layers now)
* Test the domain adaptation results.
* Add the action recognition example later when I have time.

Although I have checked them so many times to avoid mistakes, I am not very sure about whether they are right yet because it is the first time for me to push so many things. If any problem like codes, docstring format, request ways, etc., please do not hesitate to tell me! Thanks a lot!

Kind regards,
Xianyuan

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,6 +8,8 @@
 examples/data/
 examples/*/data/
 examples/*/outputs/
+examples/*/lightning_logs/
+examples/action_dann_lightn/configs/
```
Comment (haipinglu) [inline .gitignore:12]: Not sharing example config files?

File: examples/action_dann_lightn/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+`python main.py --cfg configs/EPIC-D12D2-CDAN.yaml --gpus 1`
+
+### 3. Related Kale core
+
+
+`kale.embed.video_cnn`: Video feature extractor networks (Res3D, I3D)
+
+`kale.loaddata.video_access`: Data loaders for video datasets
+
+`kale.loaddata.video_data`: `basic_video_dataset` for loading data from GTEA, KITCHEN and ADL Datasets. `epickitchen` inherited from `basic_video_dataset` for EPIC-Kitchen dataset
```
Comment (haipinglu) [inline examples/action_dann_lightn/README.md:25]: I have commented on PR from @RaivoKoot to remove the many extra data but it seems that my suggestion may not be good.
This has duplication with @RaivoKoot 's PR so we should maintain one after discussion.

File: examples/action_dann_lightn/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+# core dependencies
+pytorch==1.7.0
+pytorch-lightning==1.0.2
+numpy==1.19.1
+Python==3.8
+
+# utils
+torchvision==0.8.1
+torchsummary==1.5.1
+yacs==0.1.8
```
Comment (haipinglu) [inline examples/action_dann_lightn/requirements.txt:13]: I think I need one for digits_dann. Have you tested digits_dann and is it proper/safe to put the same requirements as action_dann?

File: kale/embed/video_cnn/i3d.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
+"""
+Define i3d on Action Recognition from https://ieeexplore.ieee.org/document/8099985
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:2]: Cap in doc? I3D?

File: kale/embed/video_cnn/i3d.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+"""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import numpy as np
+
+
+class MaxPool3dSamePadding(nn.MaxPool3d):
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:13]: Obvious and no need for docstrings?

File: kale/embed/video_cnn/i3d.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        pad_w_b = pad_w - pad_w_f
+
+        pad = (pad_w_f, pad_w_b, pad_h_f, pad_h_b, pad_t_f, pad_t_b)
+        # print x.size()
+        # print pad
+        x = F.pad(x, pad)
+        return super(MaxPool3dSamePadding, self).forward(x)
+
+
+class Unit3D(nn.Module):
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:48]: Obvious and no need for docstrings?

File: kale/embed/video_cnn/i3d.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+        x = self.conv3d(x)
+        if self._use_batch_norm:
+            x = self.bn(x)
+        if self._activation_fn is not None:
+            x = self._activation_fn(x)
+        return x
+
+
+class InceptionModule(nn.Module):
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:123]: Obvious and no need for docstrings?

File: kale/embed/video_cnn/i3d.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+    def forward(self, x):
+        b0 = self.b0(x)
+        b1 = self.b1b(self.b1a(x))
+        b2 = self.b2b(self.b2a(x))
+        b3 = self.b3b(self.b3a(x))
+        return torch.cat([b0, b1, b2, b3], dim=1)
+
+
+class InceptionI3d(nn.Module):
+    """Inception-v1 I3D architecture.
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:152]: Yes, I think this is better.

File: kale/embed/video_cnn/i3d.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+                             activation_fn=None,
+                             use_batch_norm=False,
+                             use_bias=True,
+                             name='logits')
+
+        # self.cls_rgb = Cls_rgb()
+
+        self.build()
+
+    def replace_logits(self, num_classes):
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:312]: Obvious and no need for docstrings?

File: kale/embed/video_cnn/i3d.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        self._num_classes = num_classes
+        self.logits = Unit3D(in_channels=384 + 384 + 128 + 128, output_channels=self._num_classes,
+                             kernel_shape=[1, 1, 1],
+                             padding=0,
+                             activation_fn=None,
+                             use_batch_norm=False,
+                             use_bias=True,
+                             name='logits')
+
+    def build(self):
```
Comment (haipinglu) [inline kale/embed/video_cnn/i3d.py:322]: Obvious and no need for docstrings?

File: kale/embed/video_cnn/res3d.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
+"""
+Define C3D, MC3_18, Res3D_18, Res2plus1D_18 on Action Recognition from https://arxiv.org/abs/1711.11248
+Created by Xianyuan Liu from modifying https://github.com/pytorch/vision/blob/master/torchvision/models/video/resnet.py
+"""
```
Comment (haipinglu) [inline kale/embed/video_cnn/res3d.py:4]: Check whether those w/o docstrings are obvious enough (e.g. for our new MSc who will work on AR). If not, add some docstrings.

File: kale/loaddata/video_data/epickitchen.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
+import os
+import random
+
+from .basic_video_dataset import BasicVideoDataset
+
+import pickle
+
+
+class EPIC(BasicVideoDataset):
```
Comment (haipinglu) [inline kale/loaddata/video_data/epickitchen.py:9]: Align with @RaivoKoot

**Review by haipinglu — CHANGES_REQUESTED** [review]
Great job and solid contribution. I think you know ada much better than me now. Thanks for fixing many problems in my code. You should be a reviewer. How about upload a GitHub profile photo?

We just need to discuss with @RaivoKoot about the naming convention and duplication before merging.

File: examples/action_dann_lightn/requirements.txt (hunk 3 above)
Comment (XianyuanLiu) [reply examples/action_dann_lightn/requirements.txt:13]: `digit_dann` can also run in this environment, and I will update and unify both `requirements.txt` files. But we may need to unify our Python and PyTorch version for the whole PyKale at first. Then I can upgrade or downgrade this requirement.

**Comment by XianyuanLiu** [comment]
### Changes
1. Added two config examples in the `configs/`.
2. Updated the `requirements.txt` for `action_dann_lightn` and `digits_dann_lightn`.
3. Added docstrings for `i3d.py` and `res3d.py`.
4. Updated the code format related to `action_dann_lightn` and `digits_dann_lightn` according to Pycharm advice.
5. Used the final average pooling layer output of each model as the feature extractor output.

Now we can choose R3D_18, R2PLUS1D_18, MC3_18 and I3D as the feature extractor and CDAN, CDAN-E, DANN and DAN as the DA method.

### TODO
1. Unify the video data loader with Raivo.
Renamed `video_transform.py` to `video_transform_for_action_dann.py` because of the conflict with Raivo. It is a temporary strategy before this task completed.

Kind regards,
Xianyuan

**Comment by haipinglu** [comment]
> ### Changes
> 1. Added two config examples in the `configs/`.
> 2. Updated the `requirements.txt` for `action_dann_lightn` and `digits_dann_lightn`.
> 3. Added docstrings for `i3d.py` and `res3d.py`.
> 4. Updated the code format related to `action_dann_lightn` and `digits_dann_lightn` according to Pycharm advice.
> 5. Used the final average pooling layer output of each model as the feature extractor output.
> 
> Now we can choose R3D_18, R2PLUS1D_18, MC3_18 and I3D as the feature extractor and CDAN, CDAN-E, DANN and DAN as the DA method.
> 
> ### TODO
> 1. Unify the video data loader with Raivo.
>    Renamed `video_transform.py` to `video_transform_for_action_dann.py` because of the conflict with Raivo. It is a temporary strategy before this task completed.
> 
> Kind regards,
> Xianyuan

Great to see the progress. I assume that the merge will happen after the listed todos. video_transform_for_action_dann.py is a bit long. Have you checked how pytorch name modules of similar nature? Possible to be shorter? (Not critical though, just cosmetics). We can talk on Wed.

**Comment by XianyuanLiu** [comment]
> > ### Changes
> > 
> > 1. Added two config examples in the `configs/`.
> > 2. Updated the `requirements.txt` for `action_dann_lightn` and `digits_dann_lightn`.
> > 3. Added docstrings for `i3d.py` and `res3d.py`.
> > 4. Updated the code format related to `action_dann_lightn` and `digits_dann_lightn` according to Pycharm advice.
> > 5. Used the final average pooling layer output of each model as the feature extractor output.
> > 
> > Now we can choose R3D_18, R2PLUS1D_18, MC3_18 and I3D as the feature extractor and CDAN, CDAN-E, DANN and DAN as the DA method.
> > ### TODO
> > 
> > 1. Unify the video data loader with Raivo.
> >    Renamed `video_transform.py` to `video_transform_for_action_dann.py` because of the conflict with Raivo. It is a temporary strategy before this task completed.
> > 
> > Kind regards,
> > Xianyuan
> 
> Great to see the progress. I assume that the merge will happen after the listed todos. video_transform_for_action_dann.py is a bit long. Have you checked how pytorch name modules of similar nature? Possible to be shorter? (Not critical though, just cosmetics). We can talk on Wed.
> 
> OIC. You said it is temporary so this long module name will be removed later when merging, right?

Yes, after combining with Raivo's data loader, there will be only one `video_transform.py`.

**Pull request merged**